### PR TITLE
024 registration system

### PR DESCRIPTION
**Pull request by james-walshe**
- Created a registration form page
- Created a GET and POST route 
- Created a user service and validator class with unit tests

### CODE REVIEW

File: .vscode/launch.json [hunk 1]
```diff
@@ -0,0 +1,17 @@
+{
```
Comment (andrew-fawcett) [inline .vscode/launch.json:1]: I see the folder this is in is added to your .gitignore, is this meant to be pushed?

File: app/routes.js [hunk 2]
```diff
@@ -1,33 +1,129 @@
 const express = require('express')
+const JobRoleValidator = require('./validator/JobRoleValidator')
+const UserValidator = require('./validator/UserValidator')
+const UserService = require('./service/UserService')
+const js = require('./service/JobService')
```
Comment (andrew-fawcett) [inline app/routes.js:5]: Constant for JobService is also declared in line 9, remove one of these and refactor

File: app/routes.js [hunk 3]
```diff
@@ -1,33 +1,129 @@
...
 const router = express.Router()
 
 // Add your routes here - above the module.exports line
 const JobService = require('./service/JobService');
+const e = require('express')
```
Comment (andrew-fawcett) [inline app/routes.js:10]: Constant for express is also declared in line 1, remove one of these and refactor

File: app/routes.js [hunk 4]
```diff
@@ -1,33 +1,129 @@
...
         res.locals.errormessage = "An error occured when retrieving the list of Job Roles"
     }
-    
-    res.render('jobroles', { jobroles: data } ) 
+
+    res.render('jobroles', { jobroles: data })
+});
+
+// render the addjobroles.html page 
+router.get('/addjobrole', async (req, res) => {
```
Comment (andrew-fawcett) [inline app/routes.js:26]: Lines here down to 76 are from user story 012 which hasn't been approved yet, should they be part of this story?

File: .vscode/launch.json (hunk 1 above)
Comment (james-walshe) [reply .vscode/launch.json:1]: No, shouldn't be there.

**Review by andrew-fawcett — APPROVED** [review]

File: test/unit/service/UserServiceTests.js [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+
+            var user = {
+                "email": "james@gmail.com",
+                "password": "Password12!",
+                "role": JSON.stringify({ "roleID": 1, "roleName": "Engineer" })
+            }
+
+            // mocking an error 500 response from backend
+            var mock = new MockAdapter(axios);
+            mock.onGet(UserService.REGISTRATION).reply(500);
```
Comment (shaunganley) [inline test/unit/service/UserServiceTests.js:21]: add a test for 400 being returned from API

File: test/unit/validator/UserValidatorTests.js [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+            let user = {
+                "email": "james@gmail.com",
+                "password": "",
+                "role": 1
+            }
+
+            expect(UserValidator.validateUser(user).passwordError).to.equal("Password has to be at least 8 characters, have one upper case, one lower case and one special character.")
+        })
+
+        it('should return error when password is doesnt include a capital letter', () => {
```
Comment (shaunganley) [inline test/unit/validator/UserValidatorTests.js:74]: add test for all upper case password

File: test/unit/validator/UserValidatorTests.js [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+                "role": JSON.stringify({"roleID" : 1, "roleName" :"Engineer"})
+            }
+            expect(UserValidator.validateUser(user).passwordError).to.equal(undefined)
+
+            var user = {
+                "email": "james@gmail.com",
+                "password": "Password12!",
+                "role": JSON.stringify({"roleID" : 1, "roleName" :"Engineer"})
+            }
+            expect(UserValidator.validateUser(user).roleError).to.equal(undefined)
```
Comment (shaunganley) [inline test/unit/validator/UserValidatorTests.js:30]: are these copy/paste errors?

**Review by shaunganley — CHANGES_REQUESTED** [review]

File: test/unit/validator/UserValidatorTests.js (hunk 7 above)
Comment (james-walshe) [reply test/unit/validator/UserValidatorTests.js:30]: No, the validateUser method alters the user object causing the following validateUser method to throw an exception.

Comment (shaunganley) [reply test/unit/validator/UserValidatorTests.js:30]: I think my confusion is that you define the exact same user 3 times, please create 1 user object followed by the 3 expects

**Review by shaunganley — APPROVED** [review]

**Review by shaunganley — APPROVED** [review]

**Review by shaunganley — APPROVED** [review]

**Pull request merged**